# Madhava series # 

In this noteboook, we will be using the Madhava series to find an approximation for  $\frac{\pi}{4}$. The Madhava series is a converging series so we will try see how accurate our value for $\frac{\pi}{4}$ gets as we compute more terms with a given tolerance[1]. The Madhava series is known as:

$$  
\frac{\pi}{4} =\sum_{n=0}^\infty \frac{(−1)^n}{2n+1}
$$

In [1]:
#First import our libraries

import numpy as np

#Then form our equation in a while loop


def madhava_series(tolerance):
    n = 1 
    value = 1
    newterm = tolerance*2
    
    while ((newterm**2)**.5) > tolerance:
        newterm = ((-1) ** n)/(2*n+1)
        print(f'pi/4 is {value}')
        print(f'Difference between actual value of pi/4 is {0.25*np.pi-value}') 
        value += newterm
        n += 1
    


In [2]:
pi_over_4 = madhava_series(1e-1)

pi/4 is 1
Difference between actual value of pi/4 is -0.21460183660255172
pi/4 is 0.6666666666666667
Difference between actual value of pi/4 is 0.11873149673078154
pi/4 is 0.8666666666666667
Difference between actual value of pi/4 is -0.08126850326921842
pi/4 is 0.7238095238095239
Difference between actual value of pi/4 is 0.061588639587924376
pi/4 is 0.8349206349206351
Difference between actual value of pi/4 is -0.049522471523186784


In [3]:
pi_over_4 = madhava_series(1e-2)

pi/4 is 1
Difference between actual value of pi/4 is -0.21460183660255172
pi/4 is 0.6666666666666667
Difference between actual value of pi/4 is 0.11873149673078154
pi/4 is 0.8666666666666667
Difference between actual value of pi/4 is -0.08126850326921842
pi/4 is 0.7238095238095239
Difference between actual value of pi/4 is 0.061588639587924376
pi/4 is 0.8349206349206351
Difference between actual value of pi/4 is -0.049522471523186784
pi/4 is 0.7440115440115441
Difference between actual value of pi/4 is 0.041386619385904155
pi/4 is 0.8209346209346211
Difference between actual value of pi/4 is -0.03553645753717283
pi/4 is 0.7542679542679545
Difference between actual value of pi/4 is 0.031130209129493824
pi/4 is 0.8130914836797192
Difference between actual value of pi/4 is -0.027693320282270895
pi/4 is 0.7604599047323508
Difference between actual value of pi/4 is 0.024938258665097468
pi/4 is 0.8080789523513985
Difference between actual value of pi/4 is -0.022680788953950204
pi/4 is 0.7646

We can't say our result is accurate to a tolerance of $10^{-n}$, since its the term after may change the value by rounding

In [4]:
def madhava_series2(tolerance):
    n = 1 
    value = 1
    newterm = tolerance*2
    
    while ((newterm**2)**.5) > tolerance:
        newterm = ((-1) ** n)/(2*n+1)
        print(f'pi/4 is {value}')
        print(f'The number of terms used is {n}') 
        value += newterm
        n += 1

In [5]:
%%time
pi_over_four = madhava_series2(1e-1)

pi/4 is 1
The number of terms used is 1
pi/4 is 0.6666666666666667
The number of terms used is 2
pi/4 is 0.8666666666666667
The number of terms used is 3
pi/4 is 0.7238095238095239
The number of terms used is 4
pi/4 is 0.8349206349206351
The number of terms used is 5
CPU times: total: 0 ns
Wall time: 0 ns


In [6]:
%%time
pi_over_four = madhava_series2(1e-2)

pi/4 is 1
The number of terms used is 1
pi/4 is 0.6666666666666667
The number of terms used is 2
pi/4 is 0.8666666666666667
The number of terms used is 3
pi/4 is 0.7238095238095239
The number of terms used is 4
pi/4 is 0.8349206349206351
The number of terms used is 5
pi/4 is 0.7440115440115441
The number of terms used is 6
pi/4 is 0.8209346209346211
The number of terms used is 7
pi/4 is 0.7542679542679545
The number of terms used is 8
pi/4 is 0.8130914836797192
The number of terms used is 9
pi/4 is 0.7604599047323508
The number of terms used is 10
pi/4 is 0.8080789523513985
The number of terms used is 11
pi/4 is 0.7646006914818333
The number of terms used is 12
pi/4 is 0.8046006914818333
The number of terms used is 13
pi/4 is 0.7675636544447964
The number of terms used is 14
pi/4 is 0.802046413065486
The number of terms used is 15
pi/4 is 0.769788348549357
The number of terms used is 16
pi/4 is 0.8000913788523872
The number of terms used is 17
pi/4 is 0.7715199502809587
The number of t

I can't say what the time would be for when we use a tolerance of $10^{-7}$ since just using $10^{-2}$ uses 50 terms compared to $10^{-1}$ using 5. I tried it but my computer went extremely slow and the list was way too long. Hence for part a, the values tried only were -1 and -2 since -7 would make the notebook extremely laggy when opening
I tried it with $10^{-1}$ & $10^{-2}$ and had a wall time of 0ms and 2ms respectively. Based on this, would think 16 decimal places would take about 0.06ms and 32 decimal places would be 0.12 ms. I used simple fractions to estimate these. But it's worth noting that there is a big uncertainty in the times i got, hence the times I estimated. 
Overall, I think this is a great algorithm to calculate pi since its fairly simple, given that you have a powerful enough computer to withstand extremely low tolerance values.

## References 

[1] Chislett C, Dash L and Waugh B. *PHAS0007 Unit 6: Functions and while loops: Defining a function to calculate the Madhava series
.* [online] UCL: London; 2022 [Accessed 18 November 2022]. Available from: https://moodle.ucl.ac.uk/course/view.php?id=31674